tests/s3_data_plane_tests.c

```diff
@@ -2519,6 +2520,121 @@ static int s_test_s3_default_fail_headers_callback(struct aws_allocator *allocat
     return 0;
 }
 
+static struct aws_atomic_var s_test_headers_callback_invoked;
+
+static int s_s3_test_headers_callback_check(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_http_headers *headers,
+    int response_status,
+    void *user_data) {
+    (void)meta_request;
+    (void)headers;
+    (void)response_status;
+    (void)user_data;
+
+    /* sets flag used to check if callback was invoked */
+    aws_atomic_store_int(&s_test_headers_callback_invoked, 1);
```

Comment:
Maybe increment the atomic to check the callback only invoked once?

Reply:
Great suggestion, changed to a counter and improved test to check the callback is invoked exactly once.